File: languages/swift/src/PositionTape.swift
import CryptoKit
import Foundation

public let DefaultSearchLength = 100_003

public struct Mismatch: Equatable {
    public let position: Int
    public let expected: Character?
    public let received: Character?
}

public struct ValidationResult: Equatable {
    public let isValid: Bool
    public let expectedLength: Int
    public let receivedLength: Int
    public let truncationPoint: Int?
    public let firstMismatch: Mismatch?
}

public enum PositionTape {
    private static var indexCache: [Int: [String: [Int]]] = [:]

    public static func Generate(_ length: Int) -> String {
        precondition(length >= 0, "length must be non-negative")

        var output = ""
        output.reserveCapacity(length)
        var cursor = 1

        while output.count < length {
            if cursor % 10 == 0 {
                let marker = String(cursor / 10)
                let remaining = length - output.count
                output += String(marker.prefix(remaining))
                cursor += marker.count
            } else {
                output += String(cursor % 10)
                cursor += 1
            }
        }

        return output
    }

    public static func GetMarkerCompleteLength(_ length: Int) -> Int {
        precondition(length >= 0, "length must be non-negative")

        var cursor = 1
        while cursor <= length {
            if cursor % 10 == 0 {
                let markerLength = String(cursor / 10).count
                let markerEnd = cursor + markerLength - 1
                if length < markerEnd {
                    return markerEnd
                }
                cursor += markerLength
            } else {
                cursor += 1
            }
        }

        return length
    }

    public static func GenerateMarkerComplete(_ length: Int) -> String {
        Generate(GetMarkerCompleteLength(length))
    }

    public static func Locate(_ fragment: String) -> Int {
        if fragment.isEmpty {
            return 1
        }

        let tape = Generate(DefaultSearchLength)
        guard let range = tape.range(of: fragment) else {
            return -1
        }
        return tape.distance(from: tape.startIndex, to: range.lowerBound) + 1
    }

    public static func FindFirstMismatch(_ expected: String, _ received: String) -> Mismatch? {
        let expectedChars = Array(expected)
        let receivedChars = Array(received)
        let sharedLength = min(expectedChars.count, receivedChars.count)

        for index in 0..<sharedLength {
            if expectedChars[index] != receivedChars[index] {
                return Mismatch(position: index + 1, expected: expectedChars[index], received: receivedChars[index])
            }
        }

        if expectedChars.count == receivedChars.count {
            return nil
        }

        let position = sharedLength + 1
        let expectedChar = position <= expectedChars.count ? expectedChars[position - 1] : nil
        let receivedChar = position <= receivedChars.count ? receivedChars[position - 1] : nil
        return Mismatch(position: position, expected: expectedChar, received: receivedChar)
    }

    public static func Validate(_ receivedText: String, _ expectedLength: Int) -> ValidationResult {
        let expected = Generate(expectedLength)
        let mismatch = FindFirstMismatch(expected, receivedText)
        let truncationPoint =
            mismatch != nil && receivedText.count < expectedLength && expected.hasPrefix(receivedText)
            ? receivedText.count + 1
            : nil

        return ValidationResult(
            isValid: mismatch == nil,
            expectedLength: expectedLength,
            receivedLength: receivedText.count,
            truncationPoint: truncationPoint,
            firstMismatch: mismatch
        )
    }

    public static func FindTruncationPoint(_ receivedText: String) -> Int {
        FindFirstMismatch(Generate(receivedText.count), receivedText)?.position ?? receivedText.count + 1
    }

    public static func HashFragment(_ fragment: String) -> String {
        let digest = SHA256.hash(data: Data(fragment.utf8))
        return digest.map { String(format: "%02x", $0) }.joined()
    }

    public static func BuildWindowIndex(_ windowSize: Int) -> [String: [Int]] {
        precondition(windowSize > 0, "windowSize must be positive")
        precondition(windowSize <= DefaultSearchLength, "windowSize cannot exceed the default search length")

        let tape = Array(Generate(DefaultSearchLength))
        var index: [String: [Int]] = [:]
        for offset in 0...(tape.count - windowSize) {
            let fragment = String(tape[offset..<(offset + windowSize)])
            index[HashFragment(fragment), default: []].append(offset + 1)
        }
        return index
    }

    public static func LocateByHash(_ fragmentHash: String, _ windowSize: Int) -> [Int] {
        let normalizedHash = fragmentHash.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        if indexCache[windowSize] == nil {
            indexCache[windowSize] = BuildWindowIndex(windowSize)
        }
        return indexCache[windowSize]?[normalizedHash] ?? []
    }
}

File: languages/php/src/PositionTape.php
<?php

declare(strict_types=1);

namespace PositionTape;

const DEFAULT_SEARCH_LENGTH = 100003;

final class Mismatch
{
    public function __construct(
        public int $position,
        public ?string $expected,
        public ?string $received
    ) {
    }
}

final class ValidationResult
{
    public function __construct(
        public bool $isValid,
        public int $expectedLength,
        public int $receivedLength,
        public ?int $truncationPoint,
        public ?Mismatch $firstMismatch
    ) {
    }
}

function assertNonNegativeLength(int $length): void
{
    if ($length < 0) {
        throw new \InvalidArgumentException('length must be non-negative');
    }
}

function Generate(int $length): string
{
    assertNonNegativeLength($length);
    $output = '';
    $cursor = 1;

    while (strlen($output) < $length) {
        if ($cursor % 10 === 0) {
            $marker = (string) intdiv($cursor, 10);
            $remaining = $length - strlen($output);
            $output .= substr($marker, 0, $remaining);
            $cursor += strlen($marker);
        } else {
            $output .= (string) ($cursor % 10);
            $cursor += 1;
        }
    }

    return $output;
}

function GetMarkerCompleteLength(int $length): int
{
    assertNonNegativeLength($length);
    $cursor = 1;

    while ($cursor <= $length) {
        if ($cursor % 10 === 0) {
            $markerLength = strlen((string) intdiv($cursor, 10));
            $markerEnd = $cursor + $markerLength - 1;
            if ($length < $markerEnd) {
                return $markerEnd;
            }
            $cursor += $markerLength;
        } else {
            $cursor += 1;
        }
    }

    return $length;
}

function GenerateMarkerComplete(int $length): string
{
    return Generate(GetMarkerCompleteLength($length));
}

function FindFirstMismatch(string $expected, string $received): ?Mismatch
{
    $sharedLength = min(strlen($expected), strlen($received));
    for ($index = 0; $index < $sharedLength; $index += 1) {
        if ($expected[$index] !== $received[$index]) {
            return new Mismatch($index + 1, $expected[$index], $received[$index]);
        }
    }

    if (strlen($expected) === strlen($received)) {
        return null;
    }

    $position = $sharedLength + 1;
    return new Mismatch(
        $position,
        $position <= strlen($expected) ? $expected[$position - 1] : null,
        $position <= strlen($received) ? $received[$position - 1] : null
    );
}

function Validate(string $receivedText, int $expectedLength): ValidationResult
{
    $expected = Generate($expectedLength);
    $mismatch = FindFirstMismatch($expected, $receivedText);
    $truncationPoint = null;

    if ($mismatch !== null && strlen($receivedText) < $expectedLength && str_starts_with($expected, $receivedText)) {
        $truncationPoint = strlen($receivedText) + 1;
    }

    return new ValidationResult(
        $mismatch === null,
        $expectedLength,
        strlen($receivedText),
        $truncationPoint,
        $mismatch
    );
}

function FindTruncationPoint(string $receivedText): int
{
    $mismatch = FindFirstMismatch(Generate(strlen($receivedText)), $receivedText);
    return $mismatch === null ? strlen($receivedText) + 1 : $mismatch->position;
}

function Locate(string $fragment): int
{
    if ($fragment === '') {
        return 1;
    }

    $index = strpos(Generate(DEFAULT_SEARCH_LENGTH), $fragment);
    return $index === false ? -1 : $index + 1;
}

function HashFragment(string $fragment): string
{
    return hash('sha256', $fragment);
}

function BuildWindowIndex(int $windowSize): array
{
    if ($windowSize <= 0) {
        throw new \InvalidArgumentException('windowSize must be positive');
    }
    if ($windowSize > DEFAULT_SEARCH_LENGTH) {
        throw new \InvalidArgumentException('windowSize cannot exceed the default search length');
    }

    $tape = Generate(DEFAULT_SEARCH_LENGTH);
    $index = [];
    for ($offset = 0; $offset <= strlen($tape) - $windowSize; $offset += 1) {
        $hash = HashFragment(substr($tape, $offset, $windowSize));
        $index[$hash][] = $offset + 1;
    }
    return $index;
}

function LocateByHash(string $fragmentHash, int $windowSize): array
{
    static $cache = [];
    $normalizedHash = strtolower(trim($fragmentHash));
    if (!array_key_exists($windowSize, $cache)) {
        $cache[$windowSize] = BuildWindowIndex($windowSize);
    }
    return $cache[$windowSize][$normalizedHash] ?? [];
}
